### ISSUE TITLE
Make ISession.IsDirty() method safe, effective and useful

### ISSUE THREAD

**Issue by bahusoid** [nhibernate/nhibernate-core#1413]
ISession.IsDirty() method description:
> Does this ISession contain any changes which must be synchronized with the database? Would any SQL be executed if we flushed this session?

**So expected behavior:** method should not persist any entities (work in read-only mode) and doesn't throw exceptions for dirty session.

**And current behavior:** It's very unexpected that this call can lead to exceptions or even data modifications in DB:
1) Exception is thrown when transient ManyToOne object  is present in session.
2) And data modifications - all cascaded entities are saved (which I think is very dangerous behavior)

I will try to provide failing tests today.

My use case for using this ISession.IsDirty() is the following:
```C#
var entity = LoadEntity();
//Do some modifications to entity like entity.Children.Add(child); which is mapped with Cascade.All
entity.Children.Add(CreateNewChild());

//... And at the end of the request:
if(session.IsDirty())
{
	//do some resource consuming stuff like open DTC transaction or something else that needs to be done only if database modifications are requried
	using(var dtcTransaction = OpenDTCTransaction())
	{
		session.Flush();
		//...do some stuff with legacy VB objects
	}
}

```

With current implementation I can't use `IsDirty()` as it will save child `entity.Children` elements without transaction context - very dangerous and unexpected.

If there is a valid scenario with current implementation - could you please explain it?

I understand why  cascade updates are executed - current implementation call full PreFlush life-cycle. But is it really necessary? Can't we simply check for any queued action and for any first dirty object in session?

**Comment by fredericDelaporte** [comment]
Seems related to [NH-2727](https://nhibernate.jira.com/browse/NH-2727) and [NH-2136](https://nhibernate.jira.com/browse/NH-2136) which were closed as "Not an issue" without much explanations.
About NH-2136, the case happens with identity generator. Maybe there is some good technical constraint mandating this behavior with identity, but it should then be documented on `IsDirty` (and added to the list of reasons it is not recommended to use identity for generating ids).

**Comment by hazzik** [comment]
Maybe some information is in the mailing lists (I’m away sleeping, so could not check now)

**Comment by hazzik** [comment]
Read the linked issues. I don’t know why this is an “expected behavior”, @fabiomaulo. Could you please elaborate? I agree that the behavior should be as described by @bahusoid.

**Comment by fabiomaulo** [comment]
I can't remember which was the discussion around this matter... let me check...

**Comment by fabiomaulo** [comment]
Ok... IIRC things, in these cases, are wired flush mode. In NH-2727 Ricardo said he is changing the entity during saveorupdate (probably with a listener) so the session becomes dirty. IMO as he have implemented a custom-listener to do something during saveorupdate he should change even DefaultDirtyCheckEventListener.
The second issue is easier... the usage of _Identity_ strategy for the POID is the cause... the only way to know an ID is going to DB. hmmm, I think I wrote post some years ago, maybe more than one:
http://fabiomaulo.blogspot.com.ar/2008/12/identity-never-ending-story.html
http://fabiomaulo.blogspot.com.ar/2009/02/nh210-generators-behavior-explained.html
There should be some others... _Identity_ is a very ugly beast.
BTW: Perhaps the behavior of the default listener can be changed, fixing Ricardo's issue even if we have to have extremely careful because somebody may change something, in the graph of the loaded objects, using events wired to the session lifecycle  so we may have a fake state (the IsDirty throw a false but something will be saved on flush).

**Comment by fredericDelaporte** [comment]
The point with identity is, why should `IsDirty` knows the id? If it is mandatory for checking dirtiness, so be it, but otherwise it should avoid triggering the id generation (which in the case of identity triggers the whole insert).

About NH-2727, it looks to me as boiling down to this question: is it legit for `IsDirty` to raise `SaveOrUpdate` events?

By the way this issue 1413 is indeed two issues: the many-to-one failure and the id generation triggered by `IsDirty` (which is maybe bound to the fact it raises `SaveOrUpdate`, I have not checked anything in code yet about all that).

**Comment by fabiomaulo** [comment]
Yes Federic, in general terms something called `IsDirty` should not call something called `SaveOrUpdate`... probably there is _a short way to check_ and should be re-wrote.
The dirty check is not so easy when you work with collections of objects graphs where you have to deal with the cascade behavior...anyway a dirty collection become a dirty session.

**Comment by bahusoid** [comment]
>By the way this issue 1413 is indeed two issues: the many-to-one failure and the id generation triggered by IsDirty (which is maybe bound to the fact it raises SaveOrUpdate, I have not checked anything in code yet about all that).

@fredericDelaporte  Yes. It's two different issues. Should I split it in two issues? 

I prepared a fix for "many-to-one" failure in https://github.com/nhibernate/nhibernate-core/pull/1414/commits/4febb1053bb3f86714222fc1626540a54663aede. It's not related to SaveOrUpdate call. It's related to implementation of "IEntityPersister.FindDirty" which ends up in ManyToOneType.IsDirty method. And exception is thrown inside `GetIdentifier` method for transient objects:
https://github.com/nhibernate/nhibernate-core/blob/a6f97d0423b3a57a74e736e9447ca1d0cbb91e88/src/NHibernate/Type/ManyToOneType.cs#L208

Which ends up in ForeignKeys.GetEntityIdentifierIfNotUnsaved :
https://github.com/nhibernate/nhibernate-core/blob/a6f97d0423b3a57a74e736e9447ca1d0cbb91e88/src/NHibernate/Engine/ForeignKeys.cs#L250-L265

So in my fix I explicitly check if entity is transient before calling `GetIdentifier`.
It looks like "SaveOrUpdate" triggering issue is not that easy to fix. And I would really like to see at least "many-to-one fail" issue fixed. So let me know if splitting this issue could make fix  merged quicker.

**Comment by fredericDelaporte** [comment]
Yes better split the issue.

The PR should be split too, even for related issues, at least if they do not depend on each other. Otherwise the PR will have to wait for both trouble to be fixed. The minimal "split" would be to set tests for the not fixed case as `[Ignore("Not fixed yet")]`, but it would be better to have those tests removed.

**Comment by bahusoid** [comment]
Ok. I will leave this issue for "SaveOrUpdate triggering" as most discussions here related to this case.

**Comment by hazzik** [comment]
So, it seems that this is the same as NH-2136, except it happens to anything regardless of the id generator.

**Comment by hazzik** [comment]
@bahusoid can you please update the description removing the stuff concerning #1419?

**Comment by bahusoid** [comment]
@hazzik  I tried :)

**Comment by hazzik** [comment]
This is actually `p: Major` due the fact if there is no transaction it will perform a real flush to the database, which would be considered as a manual flush by the core.

**Comment by trivalik** [comment]
Does this happen also in older NHibernate 3.x?

I am right now struggling with a huge performance gap after migrate NHibernate from 3.1 to 5.3.16/5.4.2 on calling IsDirty, Since the call is near after every change in program, it lead to a less fluent work flow.

**Comment by fredericDelaporte** [comment]
I am not sure whether this bug applies to such an old version or not, but is sounds likely to me. `IsDirty` is implemented almost like a `Flush` theoretically without the updates sent to the database. That is the reason for the bug in some special cases. That implementation has not changed much as far as I know.

About the performance loss for your application, if you could pinpoint what causes it (like profiling sql queries and seeing if there are more of them issued on IsDirty calls, or other things), that would help seeing if that is something needing a fix, and then please report it as a new issue.

About the principle of calling frequently IsDirty, that is anyway a huge source of performance loss. The basic mechanism for dirty checking in NHibernate has not changed since its first releases: the session iterates all entities it holds and compares their current state with the state they were having at load time. That is a quite heavy mechanism.
(Hibernate (Java counterpart) has some extension points allowing to use different dirty checking mechanisms, but they have not yet been ported to NHibernate.)
With such a mechanism, it is better to use dirty checking the least possible. (Actually I had never used it in applications I had worked on.)
Especially, there is no point in coding something like `if (session.IsDrity() session.Flush();`: Flush anyway perform a dirty check itself, so such a code just triggers a double dirty check when the session is dirty, without improving the performances of the "not dirty case". A variation of such an anti-pattern can also be `if (session.IsDrity() session.Commit(); else session.Rollback();`: just commit instead. There is nothing to rollback anyway if the session is not dirty, so committing instead is equivalent.

**Comment by trivalik** [comment]
I will keep that in mind. That I try to make an example in my spare time. I will post at least the profiling result in a new issue.

**Comment by fredericDelaporte** [comment]
I forgot about #2752: is the current description of the issue still relevant? I mean, may we still have entities inserted into the DB on dirty checking? Normally, with #2752, we only have at worst new entities added into the session, but not sent to the database.

**Comment by trivalik** [comment]
I can confirm that the over 1 second starts with the changes in commit 89a79d29d2d1b41f81ea12714b2bb07045fc95d8.

**Comment by bahusoid** [comment]
> I forgot about https://github.com/nhibernate/nhibernate-core/pull/2752: is the current description of the issue still relevant?

No DB modifications but session state is still modified (see https://github.com/nhibernate/nhibernate-core/pull/2752#issuecomment-841216053)
